sdk/freeport/freeport.go

```diff
@@ -64,76 +165,108 @@ func initialize() {
 // be automatically released when the application terminates.
 func alloc() (int, net.Listener) {
 	for i := 0; i < attempts; i++ {
-		block := int(rand.Int31n(int32(maxBlocks)))
+		block := int(rand.Int31n(int32(effectiveMaxBlocks)))
 		firstPort := lowPort + block*blockSize
 		ln, err := net.ListenTCP("tcp", tcpAddr("127.0.0.1", firstPort))
 		if err != nil {
 			continue
 		}
-		// log.Printf("[DEBUG] freeport: allocated port block %d (%d-%d)", block, firstPort, firstPort+blockSize-1)
+		// logf("DEBUG", "allocated port block %d (%d-%d)", block, firstPort, firstPort+blockSize-1)
 		return firstPort, ln
 	}
 	panic("freeport: cannot allocate port block")
 }
 
-func tcpAddr(ip string, port int) *net.TCPAddr {
-	return &net.TCPAddr{IP: net.ParseIP(ip), Port: port}
-}
-
-// Get wraps the Free function and panics on any failure retrieving ports.
-func Get(n int) (ports []int) {
-	ports, err := Free(n)
-	if err != nil {
-		panic(err)
-	}
-
-	return ports
-}
-
-// GetT is suitable for use when retrieving unused ports in tests. If there is
-// an error retrieving free ports, the test will be failed.
-func GetT(t testing.T, n int) (ports []int) {
-	ports, err := Free(n)
-	if err != nil {
-		t.Fatalf("Failed retrieving free port: %v", err)
-	}
-
-	return ports
-}
-
-// Free returns a list of free ports from the allocated port block. It is safe
+// Take returns a list of free ports from the allocated port block. It is safe
 // to call this method concurrently. Ports have been tested to be available on
 // 127.0.0.1 TCP but there is no guarantee that they will remain free in the
 // future.
-func Free(n int) (ports []int, err error) {
+func Take(n int) (ports []int) {
 	mu.Lock()
 	defer mu.Unlock()
 
-	if n > blockSize-1 {
-		return nil, fmt.Errorf("freeport: block size too small")
-	}
-
 	// Reserve a port block
 	once.Do(initialize)
 
-	for len(ports) < n {
-		port++
+	if n > total-1 {
+		panic(fmt.Sprintf("freeport: block size too small"))
+	}
 
-		// roll-over the port
-		if port < firstPort+1 || port >= firstPort+blockSize {
-			port = firstPort + 1
+	for len(ports) < n {
+		for freePorts.Len() == 0 {
+			if total == 0 {
+				panic("impossible; there are no actual free ports in the block anymore")
```

Comment:
If we run out of ports in the shared global pool then we can't reasonably continue to execute any tests in this test binary anymore. It's not like `retry.Run` can fix anything and failing your own test with `t.Fatalf` isn't going to help either given that the global pool is depleted and all other tests will end up short-circuit failing.

If we reach this panic then something is systemically wrong with either the entire test running plumbing in consul or something is weird with the host environment.

Reply:
I changed it up so that there is a toplevel `MustTake(int) []int` and `Take(int) ([]int, error)` now where MustTake just calls Take and panics if it gets an error.

The ergonomics are better for writing tests of the freeport package, and it clarifies the call sites a bit due to the idiomatic `Must*` prefix implying "this must succeed"